File: push/src/main/java/top/trumeet/mipush/provider/db/RegisteredApplicationDb.java
package top.trumeet.mipush.provider.db;

import static top.trumeet.common.BuildConfig.DEBUG;
import static top.trumeet.mipush.provider.DatabaseUtils.daoSession;

import android.text.TextUtils;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.greenrobot.greendao.query.QueryBuilder;

import java.util.List;

import top.trumeet.common.cache.ApplicationNameCache;
import top.trumeet.common.utils.Utils;
import top.trumeet.mipush.provider.gen.db.RegisteredApplicationDao;
import top.trumeet.mipush.provider.entities.RegisteredApplication;

/**
 * Created by Trumeet on 2017/12/23.
 */

public class RegisteredApplicationDb {

    @NonNull
    public static RegisteredApplication registerApplication(String pkg) {
        RegisteredApplication registeredApplication = getRegisteredApplication(pkg);
        if (registeredApplication == null) {
            return create(pkg);
        }
        return registeredApplication;
    }

    @Nullable
    public static RegisteredApplication getRegisteredApplication(String pkg) {
        List<RegisteredApplication> list = getList(pkg);
        if (DEBUG) {
            Log.d("RegisteredApplicationDb", "register -> existing list = " + list.toString());
        }
        if (!list.isEmpty()) {
            return list.get(0);
        }
        return null;
    }

    private static @NonNull RegisteredApplication create(String pkg) {
        // TODO: Configurable defaults; use null for optional and global options?
        RegisteredApplication registeredApplication =
                new RegisteredApplication(null
                        , pkg
                        , RegisteredApplication.Type.ASK
                        , true
                        , false
                        , false
                        , false
                        , RegisteredApplication.RegisteredType.NotRegistered
                        , ApplicationNameCache.getInstance()
                        .getAppName(Utils.getApplication(), pkg).toString()

                );
        registeredApplication.setId(insert(registeredApplication));
        return registeredApplication;
    }

    public static List<RegisteredApplication> getList(@Nullable String pkg) {
        QueryBuilder<RegisteredApplication> query = daoSession.queryBuilder(RegisteredApplication.class);
        if (!TextUtils.isEmpty(pkg)) {
            query.where(RegisteredApplicationDao.Properties.PackageName.eq(pkg));
        }
        return query.list();
    }

    public static long update(RegisteredApplication application) {
        daoSession.insertOrReplace(application);
        return application.getId();
    }

    private static long insert(RegisteredApplication application) {
        return daoSession.insert(application);
    }
}

File: push/src/main/java/top/trumeet/mipush/provider/db/RegisteredApplicationDb.kt
package top.trumeet.mipush.provider.db

import android.text.TextUtils
import com.elvishew.xlog.XLog
import com.nihility.Global
import kotlinx.coroutines.runBlocking
import top.trumeet.common.BuildConfig.DEBUG
import top.trumeet.common.utils.Utils
import top.trumeet.mipush.provider.DatabaseUtils.registeredApplicationDao
import top.trumeet.mipush.provider.entities.RegisteredApplication

/**
 * Created by Trumeet on 2017/12/23.
 */
object RegisteredApplicationDb {
    private val logger = XLog.tag("RegisteredApplicationDb").build()


    @JvmStatic
    fun registerApplication(pkg: String): RegisteredApplication {
        logger.i("registerApplication() called for: $pkg")
        val registeredApplication = getRegisteredApplication(pkg)
        return registeredApplication ?: create(pkg)
    }

    @JvmStatic
    fun getRegisteredApplication(pkg: String): RegisteredApplication? {
        val list = getList(pkg)
        if (DEBUG) {
            logger.d("register -> existing list = $list")
        }
        return list.firstOrNull()
    }

    @JvmStatic
    private fun create(pkg: String): RegisteredApplication {
        val registeredApplication = RegisteredApplication(
            id = null,
            packageName = pkg,
            type = RegisteredApplication.Type.ASK,
            notificationOnRegister = true,
            registeredType = RegisteredApplication.RegisteredType.NotRegistered,
            appName = Global.ApplicationNameCache()
                .getAppName(requireNotNull(Utils.getApplication()), pkg)
                .toString()
        )
        registeredApplication.id = insert(registeredApplication)
        return registeredApplication
    }

    @JvmStatic
    fun getList(pkg: String?): List<RegisteredApplication> = runBlocking {
        if (TextUtils.isEmpty(pkg)) {
            registeredApplicationDao.getAll()
        } else {
            val item = registeredApplicationDao.getByPackageName(pkg!!)
            if (item == null) {
                emptyList()
            } else {
                listOf(item)
            }
        }
    }

    @JvmStatic
    fun update(application: RegisteredApplication): Long = runBlocking {
        val id = registeredApplicationDao.insertOrReplace(application)
        application.id = if (application.id == null || application.id == 0L) id else application.id
        application.id ?: id
    }

    @JvmStatic
    private fun insert(application: RegisteredApplication): Long = runBlocking {
        registeredApplicationDao.insert(application)
    }
}
